Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -778,31 +779,68 @@ private static List<ClusterExemplar> computeExemplars(SGDVector[] data, Map<Inte
+                                maxInnerDist));
+                    }
                 }
-
-                // First, get the entries that will be used for cluster exemplars.
-                // The first node is polled from the tree, which has the lowest outlier score out of the remaining
-                // points assigned this cluster.
-                List<Entry<Double, Integer>> partialClusterExemplars = new ArrayList<>();
-                Stream<Integer> intStream = IntStream.range(0, numExemplarsThisCluster).boxed();
-                intStream.forEach((i) -> partialClusterExemplars.add(outlierScoreIndexTree.pollFirstEntry()));
-
-                // For each of the partial exemplars in this cluster, iterate the remaining nodes in the tree to find
-                // the maximum distance between the exemplar and the members of the cluster. The other exemplars don't
-                // need to be checked here since they won't be on the fringe of the cluster.
-                for (Entry<Double, Integer> partialClusterExemplar : partialClusterExemplars) {
-                    SGDVector features = data[partialClusterExemplar.getValue()];
-                    double maxInnerDist = Double.NEGATIVE_INFINITY;
-                    for (Entry<Double, Integer> entry : outlierScoreIndexTree.entrySet()) {
-                        double distance = dist.computeDistance(features, data[entry.getValue()]);
-                        if (distance > maxInnerDist){
-                            maxInnerDist = distance;
+                // Otherwise, the number of exemplars is greater than the size of the tree, which is most likely to occur
+                // with a contrived dataset. This indicates that there were several points in this cluster with non-unique
+                // outlier scores.
+                else {
+                    int outlierScoreTreeSize = outlierScoreIndexTree.size();
+                    List<Pair<Double, Integer>> partialClusterExemplarsList = new ArrayList<>();
+                    // When many points in the cluster have non-unique outlier scores, first we'll collect the items from
+                    // the tree.
+                    Stream<Integer> intStreamTree = IntStream.range(0, outlierScoreTreeSize).boxed();
+                    intStreamTree.forEach((i) -> {
+                        Entry<Double, Integer> treeEntry = outlierScoreIndexTree.pollFirstEntry();
+                        partialClusterExemplarsList.add(new Pair<>(treeEntry.getKey(), treeEntry.getValue()));
+                    });
+
+                    // To determine the remaining exemplars, the best thing to do is randomly sample them from all the
+                    // points in this cluster. This could introduce duplicate exemplar points, but that is safer than
+                    // reducing the number of exemplars.
+                    Random rand = new Random();
```

Comment:
We can't use an unseeded random here, it's not reproducible and isn't captured in the provenance. I don't really want to break the API to fix a bug, so maybe here you could seed the random with a fixed seed (and use SplittableRandom rather than Random, it's much higher quality), then in `main` we make that seed a config parameter and default it to the seed you chose?

Reply:
Yes, thank you for pointing this out. For `main`, the idea is that we should prepare an additional, similar PR, but with a new attribute in the `HdbscanTrainer.java` perhaps: 
`private long exemplarSeed = 12345L` annotated with `Config`
and provide this attribute to the instantiation of `SplittableRandom`?